token/core/fabtoken/v1/actions/transfer.go

```diff
@@ -0,0 +1,273 @@
+	"github.com/hyperledger-labs/fabric-token-sdk/token/token"
+)
+
+const ProtocolV1 = 1
+
+type TransferActionInput struct {
+	// identifier of token to be transferred
+	ID *token.ID
+	// Output is the input transferred
+	Input *Output
+}
+
+func (a *TransferActionInput) ToProtos() (*actions.TransferActionInput, error) {
+	var id *actions.TokenID
+	if a.ID != nil {
+		id = &actions.TokenID{
+			Id:    a.ID.TxId,
+			Index: a.ID.Index,
+		}
+	}
+	var input *actions.Token
+	if a.Input != nil {
+		input = &actions.Token{
+			Owner:    a.Input.Owner,
+			Type:     string(a.Input.Type),
+			Quantity: a.Input.Quantity,
+		}
+	}
+
+	return &actions.TransferActionInput{
+		TokenId: id,
+		Input:   input,
+	}, nil
+}
+
+func (a *TransferActionInput) FromProtos(input *actions.TransferActionInput) error {
+	if input.TokenId != nil {
+		a.ID = &token.ID{
+			TxId:  input.TokenId.Id,
+			Index: input.TokenId.Index,
+		}
+	}
+	if input.Input != nil {
+		a.Input = &Output{
+			Owner:    input.Input.Owner,
+			Type:     token.Type(input.Input.Type),
+			Quantity: input.Input.Quantity,
+		}
+	}
+	return nil
+}
+
+// TransferAction encodes a fabtoken transfer
+type TransferAction struct {
+	Inputs []*TransferActionInput
+	// outputs to be created as a result of the transfer
+	Outputs []*Output
+	// Metadata contains the transfer action's metadata
+	Metadata map[string][]byte
+}
+
+func (t *TransferAction) NumInputs() int {
+	return len(t.Inputs)
+}
+
+// NumOutputs returns the number of outputs in an TransferAction
+func (t *TransferAction) NumOutputs() int {
+	return len(t.Outputs)
+}
+
+// GetSerializedOutputs returns the serialization of the outputs in a TransferAction
+func (t *TransferAction) GetSerializedOutputs() ([][]byte, error) {
+	var res [][]byte
+	for k, output := range t.Outputs {
+		if output == nil {
+			return nil, errors.Errorf("cannot serialize transfer action outputs: nil output at index [%d]", k)
+		}
+		ser, err := output.Serialize()
+		if err != nil {
+			return nil, errors.Errorf("failed to serialize output [%d] in transfer action", k)
+		}
+		res = append(res, ser)
+	}
+	return res, nil
+}
+
+// GetOutputs returns the outputs in a TransferAction
+func (t *TransferAction) GetOutputs() []driver.Output {
+	var res []driver.Output
+	for _, output := range t.Outputs {
+		res = append(res, output)
+	}
+	return res
+}
+
+// IsRedeemAt returns true if the output at the specified index is a redeemed output
+// todo update interface to account for nil t.outputs[index]
+func (t *TransferAction) IsRedeemAt(index int) bool {
+	return t.Outputs[index].IsRedeem()
+}
+
+// IsGraphHiding returns false, indicating that fabtoken does not hide the transaction graph
+func (t *TransferAction) IsGraphHiding() bool {
+	return false
+}
+
+// SerializeOutputAt marshals the output at the specified index in TransferAction
+func (t *TransferAction) SerializeOutputAt(index int) ([]byte, error) {
+	if index >= len(t.Outputs) {
```

Comment:
is there a reason we don't want to expose only one method `GetOutputAt(index int)` and then call the methods `IsRedeem()` or `Serialize()`?

Reply:
no, good point, I can open a github issue for this because it requires a few changes, but we can do it.

Reply:
ah, yes, now I remember, it is the translator. The translator defines local interfaces for the action it can translate, `token/services/network/common/rws/translator/action.go`. If we return a complex struct, then we must have an import for that struct and I didn't want that link.